File: src/roots/embeddings.py
"""
embeddings - Vector embedding generation for semantic search.

Supports multiple embedding backends:
- Sentence Transformers (BGE, MiniLM, Qwen, etc.)
- Lite mode (n-gram hashing, zero dependencies)
"""

import hashlib
from typing import Protocol

import numpy as np


class EmbedderProtocol(Protocol):
    """Protocol for embedding implementations."""

    def embed(self, text: str) -> list[float]: ...
    def embed_batch(self, texts: list[str]) -> list[list[float]]: ...


class SentenceTransformerEmbedder:
    """Generate embeddings using sentence-transformers."""

    def __init__(self, model_name: str = "BAAI/bge-base-en-v1.5"):
        self.model_name = model_name
        self._model = None

    @property
    def model(self):
        """Lazy load the model."""
        if self._model is None:
            from sentence_transformers import SentenceTransformer

            self._model = SentenceTransformer(self.model_name, trust_remote_code=True)
        return self._model

    def embed(self, text: str) -> list[float]:
        """Generate embedding for a single text."""
        vector = self.model.encode(text, normalize_embeddings=True)
        return vector.tolist()

    def embed_batch(self, texts: list[str]) -> list[list[float]]:
        """Generate embeddings for multiple texts."""
        vectors = self.model.encode(texts, normalize_embeddings=True)
        return vectors.tolist()


class LiteEmbedder:
    """
    Lightweight embedder using character n-gram hashing.

    Not as good as neural embeddings, but:
    - Zero dependencies beyond numpy
    - Instant startup
    - Good enough for small knowledge bases
    """

    def __init__(self, dim: int = 384):
        self.dim = dim

    def embed(self, text: str) -> list[float]:
        """Generate embedding using character n-gram hashing."""
        text = text.lower().strip()
        vector = np.zeros(self.dim, dtype=np.float32)

        # Character trigrams
        for i in range(len(text) - 2):
            trigram = text[i : i + 3]
            h = int(hashlib.md5(trigram.encode()).hexdigest(), 16)
            idx = h % self.dim
            vector[idx] += 1.0

        # Word unigrams
        for word in text.split():
            h = int(hashlib.md5(word.encode()).hexdigest(), 16)
            idx = h % self.dim
            vector[idx] += 2.0  # Weight words more than trigrams

        # Normalize
        norm = np.linalg.norm(vector)
        if norm > 0:
            vector = vector / norm

        return vector.tolist()

    def embed_batch(self, texts: list[str]) -> list[list[float]]:
        """Generate embeddings for multiple texts."""
        return [self.embed(t) for t in texts]


def get_embedder(
    model_name: str | None = None,
    model_type: str = "sentence-transformers",
) -> EmbedderProtocol:
    """
    Get an embedder for the specified model.

    Args:
        model_name: Model name/path. If None, uses default BGE model.
        model_type: Either "sentence-transformers" or "lite"

    Returns:
        An embedder instance.
    """
    if model_type == "lite" or model_name == "lite":
        return LiteEmbedder()

    if model_name is None:
        model_name = "BAAI/bge-base-en-v1.5"

    try:
        # Check if sentence-transformers is available
        import sentence_transformers  # noqa: F401

        return SentenceTransformerEmbedder(model_name)
    except ImportError:
        # Fall back to lite mode
        return LiteEmbedder()


def cosine_similarity(vec_a: list[float], vec_b: list[float]) -> float:
    """Compute cosine similarity between two vectors."""
    a = np.array(vec_a)
    b = np.array(vec_b)
    norm_a = np.linalg.norm(a)
    norm_b = np.linalg.norm(b)
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return float(np.dot(a, b) / (norm_a * norm_b))

File: rust/src/embeddings.rs
use md5::{Digest, Md5};
use serde::{Deserialize, Serialize};
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;
use std::path::Path;

/// Embedding dimension for lite embedder
const LITE_DIM: usize = 384;

/// Socket path for embedding server
const SOCKET_PATH: &str = "/tmp/roots-embedder.sock";

/// Trait for embedding implementations
pub trait Embedder {
    fn embed(&self, text: &str) -> Result<Vec<f32>, String>;
    fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, String>;
}

// =============================================================================
// LiteEmbedder - N-gram hashing (pure Rust, zero deps)
// =============================================================================

/// Lightweight embedder using character n-gram hashing
pub struct LiteEmbedder {
    dim: usize,
}

impl Default for LiteEmbedder {
    fn default() -> Self {
        Self::new()
    }
}

impl LiteEmbedder {
    pub fn new() -> Self {
        Self { dim: LITE_DIM }
    }

    pub fn with_dim(dim: usize) -> Self {
        Self { dim }
    }
}

impl Embedder for LiteEmbedder {
    fn embed(&self, text: &str) -> Result<Vec<f32>, String> {
        let text = text.to_lowercase();
        let text = text.trim();
        let mut vector = vec![0.0f32; self.dim];

        // Character trigrams
        let chars: Vec<char> = text.chars().collect();
        for i in 0..chars.len().saturating_sub(2) {
            let trigram: String = chars[i..i + 3].iter().collect();
            let hash = md5_hash(&trigram);
            let idx = (hash % self.dim as u128) as usize;
            vector[idx] += 1.0;
        }

        // Word unigrams (weighted more than trigrams)
        for word in text.split_whitespace() {
            let hash = md5_hash(word);
            let idx = (hash % self.dim as u128) as usize;
            vector[idx] += 2.0;
        }

        // Normalize
        let norm: f32 = vector.iter().map(|x| x * x).sum::<f32>().sqrt();
        if norm > 0.0 {
            for v in &mut vector {
                *v /= norm;
            }
        }

        Ok(vector)
    }

    fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, String> {
        texts.iter().map(|t| self.embed(t)).collect()
    }
}

/// Compute MD5 hash and return as u128
fn md5_hash(text: &str) -> u128 {
    let mut hasher = Md5::new();
    hasher.update(text.as_bytes());
    let result = hasher.finalize();
    u128::from_be_bytes(result.into())
}

// =============================================================================
// ServerEmbedder - Unix socket client for Python embedding server
// =============================================================================

#[derive(Serialize)]
struct EmbedRequest<'a> {
    cmd: &'a str,
    text: &'a str,
}

#[derive(Serialize)]
struct EmbedBatchRequest<'a> {
    cmd: &'a str,
    texts: &'a [&'a str],
}

#[derive(Serialize)]
struct PingRequest<'a> {
    cmd: &'a str,
}

#[derive(Deserialize)]
struct EmbedResponse {
    ok: bool,
    embedding: Option<Vec<f32>>,
    error: Option<String>,
}

#[derive(Deserialize)]
struct EmbedBatchResponse {
    ok: bool,
    embeddings: Option<Vec<Vec<f32>>>,
    error: Option<String>,
}

#[derive(Deserialize)]
struct PingResponse {
    ok: bool,
    model: Option<String>,
    error: Option<String>,
}

/// Embedder that uses the Python embedding server daemon
pub struct ServerEmbedder;

impl ServerEmbedder {
    pub fn new() -> Self {
        Self
    }

    /// Check if the server is running
    pub fn is_running() -> bool {
        if !Path::new(SOCKET_PATH).exists() {
            return false;
        }

        match Self::ping() {
            Ok(_) => true,
            Err(_) => false,
        }
    }

    /// Ping the server and get the model name
    pub fn ping() -> Result<String, String> {
        let request = PingRequest { cmd: "ping" };
        let response: PingResponse = send_request(&request)?;

        if response.ok {
            Ok(response.model.unwrap_or_default())
        } else {
            Err(response.error.unwrap_or_else(|| "Unknown error".to_string()))
        }
    }

    /// Get the model the server is using
    pub fn get_model() -> Result<String, String> {
        Self::ping()
    }
}

impl Default for ServerEmbedder {
    fn default() -> Self {
        Self::new()
    }
}

impl Embedder for ServerEmbedder {
    fn embed(&self, text: &str) -> Result<Vec<f32>, String> {
        let request = EmbedRequest { cmd: "embed", text };
        let response: EmbedResponse = send_request(&request)?;

        if response.ok {
            response
                .embedding
                .ok_or_else(|| "No embedding in response".to_string())
        } else {
            Err(response.error.unwrap_or_else(|| "Unknown error".to_string()))
        }
    }

    fn embed_batch(&self, texts: &[&str]) -> Result<Vec<Vec<f32>>, String> {
        let request = EmbedBatchRequest {
            cmd: "embed_batch",
            texts,
        };
        let response: EmbedBatchResponse = send_request(&request)?;

        if response.ok {
            response
                .embeddings
                .ok_or_else(|| "No embeddings in response".to_string())
        } else {
            Err(response.error.unwrap_or_else(|| "Unknown error".to_string()))
        }
    }
}

/// Send a request to the embedding server and parse the response
fn send_request<R, T>(request: &R) -> Result<T, String>
where
    R: Serialize,
    T: for<'de> Deserialize<'de>,
{
    // Connect to socket
    let mut stream =
        UnixStream::connect(SOCKET_PATH).map_err(|e| format!("Failed to connect to server: {}", e))?;

    // Set timeout
    stream
        .set_read_timeout(Some(std::time::Duration::from_secs(60)))
        .map_err(|e| format!("Failed to set timeout: {}", e))?;

    // Send request
    let json = serde_json::to_string(request).map_err(|e| format!("Failed to serialize: {}", e))?;
    stream
        .write_all(json.as_bytes())
        .map_err(|e| format!("Failed to send: {}", e))?;

    // Shutdown write side to signal end of request
    stream
        .shutdown(std::net::Shutdown::Write)
        .map_err(|e| format!("Failed to shutdown write: {}", e))?;

    // Read response (up to 1MB)
    let mut buffer = Vec::new();
    stream
        .take(1024 * 1024)
        .read_to_end(&mut buffer)
        .map_err(|e| format!("Failed to read response: {}", e))?;

    // Parse response
    serde_json::from_slice(&buffer).map_err(|e| format!("Failed to parse response: {}", e))
}

// =============================================================================
// Cosine similarity
// =============================================================================

/// Compute cosine similarity between two vectors
pub fn cosine_similarity(vec_a: &[f32], vec_b: &[f32]) -> f64 {
    if vec_a.len() != vec_b.len() {
        return 0.0;
    }

    let dot: f32 = vec_a.iter().zip(vec_b.iter()).map(|(a, b)| a * b).sum();
    let norm_a: f32 = vec_a.iter().map(|x| x * x).sum::<f32>().sqrt();
    let norm_b: f32 = vec_b.iter().map(|x| x * x).sum::<f32>().sqrt();

    if norm_a == 0.0 || norm_b == 0.0 {
        0.0
    } else {
        (dot / (norm_a * norm_b)) as f64
    }
}

// =============================================================================
// Embedder factory
// =============================================================================

/// Get an embedder for the specified model
pub fn get_embedder(model_name: Option<&str>, model_type: &str, use_server: bool) -> Box<dyn Embedder> {
    // Lite mode
    if model_type == "lite" || model_name == Some("lite") {
        return Box::new(LiteEmbedder::new());
    }

    // Try server if requested
    if use_server {
        if ServerEmbedder::is_running() {
            if let Ok(server_model) = ServerEmbedder::get_model() {
                let requested_model = model_name.unwrap_or("BAAI/bge-base-en-v1.5");
                if server_model == requested_model {
                    return Box::new(ServerEmbedder::new());
                }
            }
        }
    }

    // Fall back to lite if server not available
    // In Rust we can't load sentence-transformers, so we always fall back to lite
    // or server. The user should start the Python server for ML embeddings.
    eprintln!(
        "Warning: Embedding server not running. Using lite embedder.\n\
         For better quality, start the server: roots server start"
    );
    Box::new(LiteEmbedder::new())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lite_embedder() {
        let embedder = LiteEmbedder::new();
        let embedding = embedder.embed("hello world").unwrap();

        assert_eq!(embedding.len(), LITE_DIM);

        // Check normalization (should be unit vector)
        let norm: f32 = embedding.iter().map(|x| x * x).sum::<f32>().sqrt();
        assert!((norm - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_cosine_similarity() {
        let a = vec![1.0, 0.0, 0.0];
        let b = vec![1.0, 0.0, 0.0];
        assert!((cosine_similarity(&a, &b) - 1.0).abs() < 0.001);

        let c = vec![0.0, 1.0, 0.0];
        assert!(cosine_similarity(&a, &c).abs() < 0.001);
    }

    #[test]
    fn test_similar_texts_have_higher_similarity() {
        let embedder = LiteEmbedder::new();

        let a = embedder.embed("the quick brown fox").unwrap();
        let b = embedder.embed("the quick brown dog").unwrap();
        let c = embedder.embed("completely different text").unwrap();

        let sim_ab = cosine_similarity(&a, &b);
        let sim_ac = cosine_similarity(&a, &c);

        // Similar texts should have higher similarity
        assert!(sim_ab > sim_ac);
    }
}
